src/utils/Cypher.ts

```diff
@@ -26,19 +35,176 @@ export default class Cypher {
     return this.configuration;
   }
 
-  public static encrypt(data: string): string {
+  public static getCrypto(): CryptoSetting {
+    if (!this.cryptoSetting) {
+      this.cryptoSetting = {
+        key: this.configuration.key
+      };
+      throw new BackendError({
+        source: Constants.CENTRAL_SERVER,
+        module: MODULE_NAME,
+        method: 'getCrypto',
+        message: 'Crypto Setting is missing'
+      });
+    }
+    return this.cryptoSetting;
+  }
+
+  public static async detectConfigurationKey(tenantID: string): Promise<boolean> {
+    let isKeyChanged = false;
+    const configCryptoKey: string = Configuration.getCryptoConfig().key;
+    const keySettings = await SettingStorage.getCryptoSettings(tenantID);
+    let cryptoSettingToSave: CryptoSetting;
+    // TODO check for migrationDone
+
+    // Check if Crypto Settings exist
+    if (keySettings) {
+      this.keySetting = keySettings;
+      // Detect new config Cypher key
+      if (keySettings.crypto?.key !== configCryptoKey) {
+        if (keySettings.crypto?.migrationDone === false) {
+          throw new BackendError({
+            source: Constants.CENTRAL_SERVER,
+            module: MODULE_NAME,
+            method: 'detectConfigurationKey',
+            message: 'Cannot change crypto key if migration not done'
+          });
+          // Call a method for completing migration?
+          // Also log error?
+        }
+        cryptoSettingToSave = {
+          formerKey: keySettings.crypto.key,
+          key: configCryptoKey,
+          migrationDone: false
+        } as CryptoSetting;
+        isKeyChanged = true;
+      }
+    } else {
+      // Create New Config Crypto Key in Tenant Settings
+      cryptoSettingToSave = {
+        key: configCryptoKey
+      } as CryptoSetting;
+    }
+
+    // Key migration of senitive data
+    this.cryptoSetting = cryptoSettingToSave ? cryptoSettingToSave : keySettings.crypto;
+
+    if (cryptoSettingToSave) {
+      const keySettingToSave = {
+        id: keySettings?.id,
+        identifier: TenantComponents.CRYPTO,
+        type: CryptoSettingsType.CRYPTO,
+        crypto: cryptoSettingToSave
+      } as KeySettings;
+      await SettingStorage.saveCryptoSettings(tenantID, keySettingToSave);
+    }
+    return isKeyChanged;
+  }
+
+  public static async setMigrationDone(tenantID:string): Promise<void> {
+    this.cryptoSetting.migrationDone = true;
+    const keySettingToSave = {
+      id: this.keySetting.id,
+      identifier: TenantComponents.CRYPTO,
+      type: CryptoSettingsType.CRYPTO,
+      crypto: this.cryptoSetting
+    } as KeySettings;
+    await SettingStorage.saveCryptoSettings(tenantID, keySettingToSave);
+  }
+
+  public static getSensitiveDatainSetting(setting: SettingDB): string[] {
```

Comment:
There's might be sensitive data elsewhere.

Reply:
elsewhere outside the settings collection?

Reply:
In the chargingstations collection for example.

Reply:
But I don't see any sensitiveData field in chargingstations so how can I migrate sensitiveData from other collections?

Reply:
I think it has been missed somehow then ... 
OCPP-S can require to save user/password for HTTP authentication to query the charging station.
Could you please have a look the soap charging station client code to confirm it has been missed?

Reply:
In ev-server/srv/client/ocpp/soap/SoapChargingStationClient.ts I do not see Cypher being used, nor any user/password. I am looking in the wrong place?

Reply:
No sensitive date outside settings.